src/gpu/intel/jit/dsl/dsl.cpp

```diff
@@ -275,114 +300,159 @@ tensor_t def(const v2::layout_t &layout, const std::string &name,
     return default_ctx().def(layout, name, value);
 }
 
-lval_t &lval_t::operator=(const expr_t &obj) {
-    assign(this->var, obj);
+lval_t &lval_t::operator=(expr_t obj) {
+    assign(this->var, std::move(obj));
     return *this;
 }
 
-expr_t let(type_t type, const std::string &name, const expr_t &value) {
-    return default_ctx().let(type, name, value);
+expr_t let(type_t type, const std::string &name, expr_t value) {
+    return default_ctx().let(type, name, std::move(value));
```

Comment:
Do we really need all these moves? I think we have a lot of simialr code without `std::move`. To me, looks worse to read.

Reply:
I could restrict this to just `if_t` and `while_t`, but otherwise these seem to be necessary. The core issue is that when templating `if_/while_` dsl functions, I ended up with:

```
stmt_t create_body();
...
if_(cond, create_body())
```
The template then matches to calling the function:

```
if_(const expr_t & cond, stmt_t stmt);
```
Clang-tidy then issues warnings for a performance-pessimizing-copy unless I implement interfaces for forwarding `stmt` into the final `if_t` ir object.

Reply:
Thanks, I see. Maybe just drop the versions without lambda? Are they used in the GEMM prototype? If someone ends up manipulating IR directly, then `append(if_t::make(...)))` is always an option.

Reply:
Part of the issue is this is used by the [labmda implementation](https://github.com/uxlfoundation/oneDNN/blob/abc61dcc0642277ce16cb319838f9828ae45af62/src/gpu/intel/jit/dsl/dsl.hpp#L169). I guess I could just create a temporary variable to work around it, but it is a fairly unexpected limitation the DSL.

Reply:
But looks like this line can be changed to `append(if_t::make(...)))`. Just in general seems like DSL API shouldn't take `stmt_t` directly, it's up to IR.